_posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc

```diff
@@ -0,0 +1,170 @@
+
+All was well... until we started deploying multiple tenants at the same time, and/or different applications. We had sized the _ocpdeploy_ container to `200 millicores` and `280 Mb` of RAM. For some applications we have around 30 tenants, that we all deploy in parallel. This meant 30 pods, which accounted for `6 cores` and `8 Gb` of RAM. This seems livable, but deployments tend to be done after business hours, between 7 and 8pm for many. And we started being afraid of the impact _ocpdeploy_ itself would have if we were running multiple deployments in parallel, on top of our SpringBoot or WildFly applications, which take their toll on the cluster at startup.
+
+== Enter the Quarkus Universe
+
+So less than a year ago we decided to launch a Quarkus POC on _ocpdeploy_, the goal being to deploy it as a GraalVM executable.
+The main challenges we ran into were related to the lack of support for GraalVM in some of the libraries we were using: FreeMarker, SMBJ, the Fabric8 OpenShift client (when instantiating OpenShift specific CRDs), and a home made java agent that ships logs to a proprietary centralized logging system.
+
+Fortunately, other bits were already there in Quarkus, such as support for command mode applications, a rest client, the Vault extension, and support for the Fabric8 Kubernetes Client, which provided a nice base for a new OpenShift Client extension that was added to the core. The Kubernetes Client extension alone was a huge push for the project because when we started we did not have Argo available, so we had to implement _apply_ and _prune_ ourselves.
+
+For FreeMarker we pushed a new extension to the Quarkiverse, with code largely inspired from work by _ppalaga_ and _carlosthe19916_.
+For SMBJ we created an extension in our internal Quarkiverse. And for our logging client, I was able to draw some inspiration from the `Quarkus-logging-gelf` extension, and created an additional internal extension.
+
+Getting a library to run in native was facilitated by the very good documentation that comes from the Quarkus project and support in the framework (e.g. _processors_ and _recorders_): https://Quarkus.io/guides/writing-native-applications-tips#native-in-extension[Supporting native in a Quarkus extension]
+
+When things got a little hairy (SMBJ was trickier than we thought), we got some help from the https://medium.com/GraalVM/introducingthe-tracing-agent-simplifying-GraalVM-native-image-configuration-c3b56c486271[GraalVM tracing agent].
+
+Here is the list of extensions we are currently using: `cdi, config-yaml, FreeMarker, hibernate-validator, kubernetes-client, openshift-client, rest-client, restclient-jackson, vault` (plus internal extensions for SMBJ and our internal centralized logging system).
```

Comment:
```suggestion
Here is the list of extensions we are currently using: `cdi, config-yaml, freemarker, hibernate-validator, kubernetes-client, openshift-client, rest-client, restclient-jackson, vault` (plus internal extensions for SMBJ and our internal centralized logging system).
```

Reply:
I actually respected the case of the product as stated in https://freemarker.apache.org/ 

> What is Apache FreeMarker™?

Reply:
you are listing all the other extensions using their extension names - the others have similar rules when referred to by product name. I just suggest you use one approach not mix'em.

Reply:
ah sorry. you are completely right. this is a bad replace all that I did to make sure that all "FreeMarker" had the right case. this one should not have been replaced. good catch. thanks.